esp-hal/src/gpio/asynch.rs

```diff
@@ -0,0 +1,223 @@
+use core::{
+    sync::atomic::Ordering,
+    task::{Context, Poll},
+};
+
+use procmacros::ram;
+
+use crate::{
+    asynch::AtomicWaker,
+    gpio::{Event, Flex, GpioBank, Input, NUM_PINS},
+};
+
+#[ram]
+pub(super) static PIN_WAKERS: [AtomicWaker; NUM_PINS] = [const { AtomicWaker::new() }; NUM_PINS];
+
+impl Flex<'_> {
+    /// Wait until the pin experiences a particular [`Event`].
+    ///
+    /// The GPIO driver will disable listening for the event once it occurs,
+    /// or if the `Future` is dropped - which also means this method is **not**
+    /// cancellation-safe, it will always wait for a future event.
+    ///
+    /// Note that calling this function will overwrite previous
+    /// [`listen`][Self::listen] operations for this pin.
+    #[inline]
+    #[instability::unstable]
+    pub async fn wait_for(&mut self, event: Event) {
+        // Make sure this pin is not being processed by an interrupt handler.
+        if self.is_listening() {
+            // An interrupt may happen at any point, and the interrupt handler disables the
+            // interrupt enable bit (meaning any unlisten call here might be a race
+            // condition without the critical section). On dual cores, this may happen in
+            // parallel, with the interrupt handler running on the other core.
+            // The safe way to handle this case is to take a GPIO-global
+            // critical section so that we won't run in parallel with the
+            // interrupt handler.
+            // Note that the critical section is taken inside `unlisten`.
+
+            // The pin is already listening, let's turn that off to prevent the interrupt
+            // handler from being called while we are setting up the pin.
+            self.unlisten();
+            self.clear_interrupt();
```

Comment:
Shouldn't interrupt clearing be done unconditionally? Regardless of whether it was listening or not.

This is needed to uphold the [e-hal contract](https://github.com/rust-embedded/embedded-hal/blob/520945278942c301433c391f63a075227d9e7c84/embedded-hal-async/src/digital.rs#L172) I think.

Reply:
Actually it was already done in `listen` but maybe not entirely correctly... I added a test to verify this at least in the basic case.